ocaml/xapi/xapi_ha_vm_failover.ml

```diff
@@ -175,60 +193,452 @@ let order_f (_, vm_rec) =
+        (Db.VM.get_name_label ~__context ~self:vm)
+        (gib_of_bytes size)
+    )
+    vms_increasing
+
+let host_free_memory ~__context ~host =
+  Memory_check.host_compute_free_memory_with_maximum_compression ~__context
+    ~host None
+
+let vm_memory ~__context snapshot =
+  let policy =
+    match Helpers.check_domain_type snapshot.API.vM_domain_type with
+    | `hvm | `pv ->
+        Memory_check.Dynamic_min
+    | `pv_in_pvh | `pvh ->
+        Memory_check.Static_max
   in
-  let vms =
-    List.map
-      (fun (vm, snapshot) ->
-        let policy =
-          match Helpers.check_domain_type snapshot.API.vM_domain_type with
-          | `hvm | `pv ->
-              Memory_check.Dynamic_min
-          | `pv_in_pvh | `pvh ->
-              Memory_check.Static_max
-        in
-        (vm, total_memory_of_vm ~__context policy snapshot)
-      )
-      vms_and_snapshots
+  total_memory_of_vm ~__context policy snapshot
+
+(** Generate 2 layers Map: the outer layer is a Map from `VM_group to ranked_hosts, while
+    each ranked_hosts is a Map from vm_count to sorted host(`host, free_memory_size) List
+    in increasing order in terms of the host's free memory, where every host in the list
+    has the same number of running anti-affinity VMs resident on it: vm_count, for example:
+    group: grp1
+      vm_count: 1, hosts: [(master 500),(slave2 508)]
+    group: grp2
+      vm_count: 0, hosts: [(slave2 508)]
+      vm_count: 1, hosts: [(master 500)] *)
+let anti_affinity_grp_ranked_hosts ~__context hosts =
+  let host_to_size_map =
+    List.fold_left
+      (fun m (h, h_size) -> HostMap.add h h_size m)
+      HostMap.empty hosts
   in
+  Db.VM_group.get_all ~__context
+  |> List.filter (fun group ->
+         Db.VM_group.get_placement ~__context ~self:group = `anti_affinity
+     )
+  |> List.fold_left
+       (fun grp_to_ranked_hosts_map group ->
+         VMGroupMap.add group
+           (Xapi_vm_helpers.rank_hosts_by_vm_cnt_in_group ~__context group
+              (hosts |> List.map fst)
+           |> List.fold_left
+                (fun vm_cnt_to_hosts_map hs ->
+                  match hs with
+                  | (_h, vm_cnt) :: _ ->
+                      IntMap.add vm_cnt
+                        (hs
+                        |> List.map (fun (h, _vm_cnt) ->
+                               (h, HostMap.find h host_to_size_map)
+                           )
+                        |> List.sort bigger_than
+                        )
+                        vm_cnt_to_hosts_map
+                  | [] ->
+                      vm_cnt_to_hosts_map
+                )
+                IntMap.empty
+           )
+           grp_to_ranked_hosts_map
+       )
+       VMGroupMap.empty
+
+(** Update the 2 layers Map - grp_ranked_hosts when a VM, its memory size is vm_size and
+    it is in anti-affinity group, will be planned to run on the host. *)
+let update_anti_affinity_grp_ranked_hosts vm_size group host grp_ranked_hosts =
```

Comment:
The function is even larger than before. 😟 
Wouldn't the difficulty of updating on such a data structure (2-layer map) imply that the data structure is not good?

Reply:
I think updating the 2 layers map is not difficult. :)
```
I updated the comments as:
(** Generate 2 layers Map: the outer layer is a Map from `VM_group to ranked_hosts, while
    each ranked_hosts is a Map from vm_count to sorted host(`host, free_memory_size) List
    in increasing order in terms of the host's free memory, where every host in the list
    has the same number of running anti-affinity VMs resident on it: vm_count, for example:
    { grp1 ->
       { 0 -> [ (master, 500), (slave2, 508) ] }
    , grp2 ->
       { 0 -> [ (slave2, 508) ]
       , 1 -> [ (master, 500) ]
       }
    }
    indicates that there are 2 VM groups. For grp1, all the 2 hosts has 0 running VM
    from grp1, where master has 500 free memory, slave2 has 508 free memory, the 2 hosts are
    in increasing order in terms of the host's free memory. For grp2, slave2 has 0 running
    VM from grp2, and master has 1 running VM from grp2. *)
```
Wish it will be helpful to understanding the data structure.

Reply:
The `compare` could be a two-dimension key (vm_cnt, free_mem). Then the data structure will be a sorted list still. Creating and updating the list would be easier?

Reply:
Thanks for this excellent idea!